### PR TITLE
Add support for user-supplied file formats #56, #177

### LINKED ISSUE #56: Allow Configurable File types
Currently, allowed file types (.java, .adoc)are hardcoded, we should allow users to configure what file types he wants to allow

### PR DESCRIPTION
**Pull request by yong24s**
Fixes #56, #177

```
File formats .java and .adoc are hardcoded in the git log command of
RepoSense.

As the file formats are not changable, this means RepoSense may only
be useful for users who work on repository that consisting mainly of
java or adoc.

Let's allow an optional file formats to be passed in as a parameter so
that user could use RepoSense on repositories with different codebases.
```

### CODE REVIEW

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 1]
```diff
@@ -18,6 +18,7 @@
     private String displayName;
     private Date sinceDate;
     private Date untilDate;
+    private transient List<String> formats;
```
Comment (yamidark) [inline src/main/java/reposense/model/RepoConfiguration.java]: put this with the rest of the transient data fields

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 2]
```diff
@@ -42,6 +43,12 @@ public static void setDatesToRepoConfigs(
         }
     }
 
+    public static void setFormatsToRepoConfigs(List<RepoConfiguration> configs, List<String> formats) {
```
Comment (yamidark) [inline src/main/java/reposense/model/RepoConfiguration.java]: JavaDocs

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 3]
```diff
@@ -42,6 +43,12 @@ public static void setDatesToRepoConfigs(
...
+        for (RepoConfiguration config : configs) {
+            config.setFormats(formats);
+        }
```
Comment (yamidark) [inline src/main/java/reposense/model/RepoConfiguration.java]: `configs.forEach(config -> config.setFormats(formats));`

File: src/main/java/reposense/parser/ArgsParser.java [hunk 4]
```diff
@@ -19,6 +22,7 @@
  */
 public class ArgsParser {
     public static final String DEFAULT_REPORT_NAME = "reposense-report";
+    public static final List<String> DEFAULT_FORMATS =  Stream.of("java", "adoc").collect(Collectors.toList());
```
Comment (yamidark) [inline src/main/java/reposense/parser/ArgsParser.java]: `Collections.list("java", "adoc")`;

File: src/test/java/reposense/commits/CommitInfoExtractorTest.java [hunk 5]
```diff
@@ -23,7 +23,6 @@ public void withContentTest() {
     public void withoutContentTest() {
         Date sinceDate = TestUtil.getDate(2050, Calendar.JANUARY, 1);
         config.setSinceDate(sinceDate);
-
```
Comment (yamidark) [inline src/test/java/reposense/commits/CommitInfoExtractorTest.java]: Don't do this cleaning up on files which you aren't touching it in this PR

File: src/test/java/reposense/parser/ArgsParserTest.java [hunk 6]
```diff
@@ -41,12 +44,16 @@ public void parse_allCorrectInputs_success() throws ParseException, IOException
         Date expectedUntilDate = TestUtil.getDate(2017, Calendar.NOVEMBER, 30);
         Assert.assertEquals(expectedSinceDate, cliArguments.getSinceDate().get());
         Assert.assertEquals(expectedUntilDate, cliArguments.getUntilDate().get());
+
+        List<String> expectedFormats = Stream.concat(
+                ArgsParser.DEFAULT_FORMATS.stream(), Stream.of("html", "css", "js")).collect(Collectors.toList());
```
Comment (yamidark) [inline src/test/java/reposense/parser/ArgsParserTest.java]: The default formats won't be added in if we provide some other format and without them right? So don't do this concatenation and instead just provide them all straight.
Also, `Collections.list(...)` as shown above earlier.

File: src/test/java/reposense/parser/ArgsParserTest.java [hunk 7]
```diff
@@ -56,6 +63,10 @@ public void parse_withExtraWhitespaces_success() throws ParseException, IOExcept
...
```
Comment (yamidark) [inline src/test/java/reposense/parser/ArgsParserTest.java]: Same as above.

File: src/test/java/reposense/system/CommandRunnerTest.java [hunk 8]
```diff
@@ -31,18 +32,20 @@ public void checkoutTest() {
 
     @Test
     public void logWithContentTest() {
-        String content = CommandRunner.gitLog(TestConstants.LOCAL_TEST_REPO_ADDRESS, null, null);
+        String content =
+                CommandRunner.gitLog(TestConstants.LOCAL_TEST_REPO_ADDRESS, null, null, Collections.emptyList());
```
Comment (yamidark) [inline src/test/java/reposense/system/CommandRunnerTest.java]: Should be using the `DEFAULT_FORMATS` here (and tests below), since an empty list for us here will mean `git log` will log the commits for all file types (which they tracked). 
We should also add tests for different formats given, like `*.py` will give empty content since we have no `.py` files. 
Remember to update the test name to the 3 point format.

File: src/test/java/reposense/system/CommandRunnerTest.java [hunk 9]
```diff
@@ -52,8 +55,8 @@ public void blameRawTest() {
         Date untilDate = TestUtil.getDate(2050, Calendar.JANUARY, 1);
 
         // supply both sinceDate and untilDate -> success
-        String content = CommandRunner.blameRaw(
-                TestConstants.LOCAL_TEST_REPO_ADDRESS, "blameTest.java", sinceDate, untilDate);
+        String content =
+                CommandRunner.blameRaw(TestConstants.LOCAL_TEST_REPO_ADDRESS, "blameTest.java", sinceDate, untilDate);
```
Comment (yamidark) [inline src/test/java/reposense/system/CommandRunnerTest.java]: Unnecessary change

File: src/main/java/reposense/parser/ArgsParser.java [hunk 10]
```diff
@@ -61,6 +65,13 @@ private static ArgumentParser getArgumentParser() {
                 .setDefault(Optional.empty())
                 .help("The date to stop filtering.");
 
+        parser.addArgument("-formats")
+                .nargs("*")
+                .metavar("FORMAT")
+                .setDefault(DEFAULT_FORMATS)
+                .help("The file formats to process. "
```
Comment (yamidark) [inline src/main/java/reposense/parser/ArgsParser.java]: Are there are restrictions for file extension format? Like must it be only alphanumeric or any alphabetic? Can we add this specification here?

File: src/main/java/reposense/parser/ArgsParser.java [hunk 11]
```diff
@@ -61,6 +65,13 @@ private static ArgumentParser getArgumentParser() {
...
+                        + "If not provided, java and adoc will be used.");
```
Comment (yamidark) [inline src/main/java/reposense/parser/ArgsParser.java]: Put quotes on the `java` and `adoc` to be more specific those are the file format.

File: src/main/java/reposense/system/CommandRunner.java [hunk 12]
```diff
@@ -17,16 +18,27 @@
...
     private static boolean isWindows = isWindows();
 
-    public static String gitLog(String root, Date sinceDate, Date untilDate) {
+    public static String gitLog(String root, Date sinceDate, Date untilDate, List<String> formats) {
         Path rootPath = Paths.get(root);
 
         String command = "git log --no-merges ";
         command += getGitDateRangeArgs(sinceDate, untilDate);
-        command += " --pretty=format:\"%h|%aN|%ad|%s\" --date=iso --shortstat -- \"*.java\" -- \"*.adoc\"";
+        command += " --pretty=format:\"%h|%aN|%ad|%s\" --date=iso --shortstat" + buildFormatsCommand(formats);
```
Comment (yamidark) [inline src/main/java/reposense/system/CommandRunner.java]: separate the `buildFormatsCommand(formats)` to a newline.

File: src/main/java/reposense/system/CommandRunner.java [hunk 13]
```diff
@@ -17,16 +18,27 @@
...
 
         return runCommand(rootPath, command);
     }
 
+    private static String buildFormatsCommand(List<String> formats) {
+        StringBuilder stringBuilder = new StringBuilder();
+        final String cmdFormat = " -- \"*.%s\"";
```
Comment (yamidark) [inline src/main/java/reposense/system/CommandRunner.java]: Can use the `addQuotes(String)` helper method below

File: src/main/java/reposense/system/CommandRunner.java [hunk 14]
```diff
@@ -17,16 +18,27 @@
...
+
+        for (String format : formats) {
+            stringBuilder.append(String.format(cmdFormat, format));
+        }
```
Comment (yamidark) [inline src/main/java/reposense/system/CommandRunner.java]: `formats.forEach(format -> stringBuilder.append(String.format(cmdFormat, format)));`
or does:
```Java
        formats.stream()
                .map(format -> String.format(cmdFormat, format))
                .forEach(stringBuilder::append);
```
seem better?

**Comment by yamidark** [comment]
https://github.com/reposense/RepoSense/blob/f758253fdba88fd3fecd00d082bba202faaa0025/src/main/java/reposense/authorship/FileInfoExtractor.java#L61-L63
Need to update here as well

**Comment by yong24s** [comment]
@eugenepeh For your review and comment on @yamidark suggestion, please.

This PR extends on #177, so that user can specify a list of file formats with
```-formats java adocs md ...```.

Therefore, I have renamed ```fileFormats``` to ```formats``` for consistency with ```-formats```.

Also please target this PR for early 1.2 release.

File: src/main/java/reposense/parser/ArgsParser.java [hunk 15]
```diff
@@ -61,6 +65,14 @@ private static ArgumentParser getArgumentParser() {
                 .setDefault(Optional.empty())
                 .help("The date to stop filtering.");
 
+        parser.addArgument("-formats")
+                .nargs("*")
+                .metavar("FORMAT")
+                .type(new AlphanumericArgumentType())
+                .setDefault(DEFAULT_FORMATS)
+                .help("The alphanumeric file formats to process. "
```
[the remark this answers is not in the document]
Comment (yong24s) [reply src/main/java/reposense/parser/ArgsParser.java]: The sanity checker only allows [A-Za-z0-9]. 
i.e. ```.java``` will fail. So I empathized the need to use ```alphanumeric```.

File: src/main/java/reposense/parser/ArgsParser.java [hunk 16]
```diff
@@ -61,6 +65,14 @@ private static ArgumentParser getArgumentParser() {
...
+                        + "If not provided, popular programming file formats will be used.");
```
[the remark this answers is not in the document]
Comment (yong24s) [reply src/main/java/reposense/parser/ArgsParser.java]: Actually both ```popular``` and ```common``` are subjective.
Let's also add the actual list of default file formats into the user guide.

[the remark this answers is not in the document]
Comment (eugenepeh) [reply src/main/java/reposense/parser/ArgsParser.java]: How about, default file formats will be used, please refer to userguide for more information.

File: docs/UserGuide.md [hunk 17]
```diff
@@ -39,11 +40,11 @@ The contribution calculation is based on the daily commits made within 00:00 to
 ### Other option:
 1. Clone this repository (or [download as zip](https://github.com/reposense/RepoSense/archive/master.zip))
 1. Execute the following command on the OS terminal inside the project directory.<br>
-Usage: `gradlew run -Dargs="-config CSV_CONFIG_FILE_PATH [-output OUTPUT_DIRECTORY] [-since DD/MM/YYYY] [-until DD/MM/YYYY]"` <br>
+Usage: `gradlew run -Dargs="-config CSV_CONFIG_FILE_PATH [-output OUTPUT_DIRECTORY] [-since DD/MM/YYYY] [-until DD/MM/YYYY] [-formats [FORMAT [FORMAT ...]]]"` <br>
```
Comment (eugenepeh) [inline docs/UserGuide.md]: [FORMAT [FORMAT ...]] ??? 

why not just `FORMAT...`

File: src/functional/java/reposense/Entry.java [hunk 18]
```diff
@@ -63,8 +63,8 @@ private void generateReport(String inputDates) throws IOException, URISyntaxExce
 
         CliArguments cliArguments = ArgsParser.parse(translateCommandline(input));
         List<RepoConfiguration> configs = CsvParser.parse(cliArguments.getConfigFilePath());
-        RepoConfiguration.setFormatsToRepoConfigs(configs, TESTING_FILE_FORMATS);
         RepoConfiguration.setDatesToRepoConfigs(configs, cliArguments.getSinceDate(), cliArguments.getUntilDate());
+        RepoConfiguration.setFormatsToRepoConfigs(configs, TESTING_FILE_FORMATS);
```
Comment (eugenepeh) [inline src/functional/java/reposense/Entry.java]: is there a need for this change?

File: src/main/java/reposense/model/CliArguments.java [hunk 19]
```diff
@@ -12,12 +13,15 @@
     private Path outputFilePath;
     private Optional<Date> sinceDate;
     private Optional<Date> untilDate;
+    private List<String> formats;
 
-    public CliArguments(Path configFilePath, Path outputFilePath, Optional<Date> sinceDate, Optional<Date> untilDate) {
+    public CliArguments(Path configFilePath, Path outputFilePath,
+                        Optional<Date> sinceDate, Optional<Date> untilDate, List<String> formats) {
```
Comment (eugenepeh) [inline src/main/java/reposense/model/CliArguments.java]: line wrap 8 spaces

**Comment by yong24s** [comment]
I believe `git log` does not explicitly state whether it is a `file format` or a `file extension`

Why I chose to use `format` was `-formats` are shorter than typing `-extensions` and `formats` are generally used for file types while `extensions` are used to refer to a property. However, both may be used interchangeably.

image

Apart from this, in the future, we may want to support `shorthands` for `CLI`. So `-formats` can be shorten to `-f`.

File: docs/UserGuide.md [hunk 20]
```diff
@@ -39,11 +40,11 @@ The contribution calculation is based on the daily commits made within 00:00 to
 ### Other option:
 1. Clone this repository (or [download as zip](https://github.com/reposense/RepoSense/archive/master.zip))
 1. Execute the following command on the OS terminal inside the project directory.<br>
-Usage: `gradlew run -Dargs="-config CSV_CONFIG_FILE_PATH [-output OUTPUT_DIRECTORY] [-since DD/MM/YYYY] [-until DD/MM/YYYY]"` <br>
+Usage: `gradlew run -Dargs="-config CSV_CONFIG_FILE_PATH [-output OUTPUT_DIRECTORY] [-since DD/MM/YYYY] [-until DD/MM/YYYY] [-formats FORMAT]"` <br>
```
Comment (eugenepeh) [inline docs/UserGuide.md]: FORMAT -> FORMAT...

**Comment by yong24s** [comment]
@damithc Prof. for your review, please.

File: src/main/java/reposense/parser/AlphanumericArgumentType.java [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+
+import java.util.regex.Pattern;
+
+import net.sourceforge.argparse4j.inf.Argument;
+import net.sourceforge.argparse4j.inf.ArgumentParser;
+import net.sourceforge.argparse4j.inf.ArgumentParserException;
+import net.sourceforge.argparse4j.inf.ArgumentType;
+
+/**
+ * Verifies a String is alphanumeric.
```
Comment (damithc) [inline src/main/java/reposense/parser/AlphanumericArgumentType.java]: A class usually _represents_ something rather than _does_ something. This class represents a Alphanumeric Argument Type
